File: backend/python/pcb_to_netlist/7_map_net_names.py
# file: 7_map_net_names.py
# Replace NET IDs with names and sort

import sys # argv, exit

def main():
    if len(sys.argv) != 4: sys.exit(1)
    pcb, nets_f, out_f = sys.argv[1:4]
    nmap = {}
    with open(pcb, 'r') as f:
        for l in f:
            if not l.startswith('RECORD=Net'): continue
            p = {x.split('=')[0]: x.split('=')[1] for x in l.split('|') if '=' in x}
            if 'ID' in p and 'NAME' in p: nmap[p['ID']] = p['NAME'].strip()
    res = []
    with open(nets_f, 'r') as f_in:
        for l in f_in:
            if not l.startswith('NET'): continue
            nid_p, conns = l.strip().split(':', 1)
            res.append(f"{nmap.get(nid_p[3:], nid_p[3:])}:{conns}")
    res.sort()
    with open(out_f, 'w') as f: f.write('\n'.join(res))

if __name__ == "__main__":
    main()

File: src/typescript/pcb_to_netlist/7_map_net_names.ts
import fs from 'fs';

export function mapNetNames(pcbPath: string, netsPath: string, outPath: string): void {
    const nmap: Record<string, string> = {};
    const pcbContent = fs.readFileSync(pcbPath, 'utf8');
    
    for (const line of pcbContent.split(/\r?\n/)) {
        if (!line.startsWith('RECORD=Net')) continue;
        const p: Record<string, string> = {};
        line.split('|').forEach(x => {
            if (x.includes('=')) {
                const [k, v] = x.split('=', 2);
                p[k] = v;
            }
        });
        if (p['ID'] && p['NAME']) {
            nmap[p['ID']] = p['NAME'].trim();
        }
    }
    
    const netsLines = fs.readFileSync(netsPath, 'utf8').split(/\r?\n/);
    const res: string[] = [];
    
    for (const line of netsLines) {
        if (!line.startsWith('NET')) continue;
        const [nid_p, conns] = line.split(':', 2);
        const nid = nid_p.substring(3);
        res.push(`${nmap[nid] || nid}:${conns}`);
    }
    
    res.sort();
    fs.writeFileSync(outPath, res.join('\n'));
}
